kv: remove dead code in `Replica.raftTruncatedStateLocked`

nvb opened the pull request:
`r.mu.state.TruncatedState` is never nil (this is already assumed in `raftTruncatorReplica.getTruncatedState`), so the logic in `Replica.raftTruncatedStateLocked` that fell back to reading from the engine and lazily populating `r.mu.state.TruncatedState` was dead code.

nvb committed:
kv: remove dead code in `Replica.raftTruncatedStateLocked`

`r.mu.state.TruncatedState` is never nil (this is already assumed in
`raftTruncatorReplica.getTruncatedState`), so the logic in
`Replica.raftTruncatedStateLocked` that fell back to reading from the
engine and lazily populating `r.mu.state.TruncatedState` was dead code.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/81879)
<!-- Reviewable:end -->

nvb committed:
kv: lock r.mu for reading around FirstIndex and LastIndex

This change weakens some of the locking around `replicaRaftStorage`'s FirstIndex
and LastIndex methods. These methods themselves don't modify state, so they only
need to hold r.mu in read mode. Making this change allows us to simplify some
upstream code.

nvb committed:
kv: remove error return values from GetFirstIndex and GetLastIndex

These methods never returned errors, so there was no reason to force all
callers to handle errors.

tbg reviewed (approved):
> Reviewed 1 of 1 files at r1, 7 of 7 files at r2, 15 of 15 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/81879)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten)

___
*[`pkg/kv/kvserver/replica_raftstorage.go` line 341 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/81879#-N3JzzzA8bdsYT5WhaG5:-N3JzzzA8bdsYT5WhaG6:b-vjkjl3) ([raw file](https://github.com/cockroachdb/cockroach/blob/5cc0d2529774783854416458e091313f975c06a1/pkg/kv/kvserver/replica_raftstorage.go#L341)):*
> ```Go
> // raftFirstIndexRLocked requires that r.mu is held for reading.
> func (r *Replica) raftFirstIndexRLocked() (uint64, error) {
> 	return (*replicaRaftStorage)(r).FirstIndex()
> ```

FirstIndex is on `raft.Storage` and we say that `r.mu` needs to be locked:

https://github.com/cockroachdb/cockroach/blob/d5b3fe7c91e1da9745b337e66ad26c776fde0aee/pkg/kv/kvserver/replica_raftstorage.go#L45-L58

Now that this is no longer true, we need to be more careful about the semantics. We must say what mutex needs to be held on a method-per-method basis. Looking at the interface, everything seems read-only superficially, though it's not true due to populating the raft entry cache. So probably best to point out in the above comment that FirstIndex only needs the read lock to be held, and the rest the exclusive lock, though it may be possible to weaken that in some cases.


<!-- Sent from Reviewable.io -->

nvb reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/81879)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @tbg)

___
*[`pkg/kv/kvserver/replica_raftstorage.go` line 341 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/81879#-N3JzzzA8bdsYT5WhaG5:-N3uPOH59bveZ3r7ABrD:b-lfyqcn) ([raw file](https://github.com/cockroachdb/cockroach/blob/5cc0d2529774783854416458e091313f975c06a1/pkg/kv/kvserver/replica_raftstorage.go#L341)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

FirstIndex is on `raft.Storage` and we say that `r.mu` needs to be locked:

https://github.com/cockroachdb/cockroach/blob/d5b3fe7c91e1da9745b337e66ad26c776fde0aee/pkg/kv/kvserver/replica_raftstorage.go#L45-L58

Now that this is no longer true, we need to be more careful about the semantics. We must say what mutex needs to be held on a method-per-method basis. Looking at the interface, everything seems read-only superficially, though it's not true due to populating the raft entry cache. So probably best to point out in the above comment that FirstIndex only needs the read lock to be held, and the rest the exclusive lock, though it may be possible to weaken that in some cases.
</blockquote></details>

Good point. I've added commentary about this here and on each of the methods in `raft.Storage`.

Doing so revealed two subtle issues that each came close to materializing as bugs. This first is resolved in https://github.com/cockroachdb/cockroach/pull/82479. The second (resolved here) is that we were stating that `replicaRaftStorage.Term` only required read access when it actually requires write access. Luckily, we weren't calling it concurrently.


<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
TFTR!

bors r=tbg
